rust-enjoyer/src/main.rs

```diff
@@ -0,0 +1,116 @@
+#[cfg(target_feature = "avx2")]
+use rust_enjoyer::sponges_avx::SpongesAvx;
+
+use rust_enjoyer::*;
+
+use rayon::prelude::*;
+use std::env;
+use std::process;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+
+fn main() {
+    #[cfg(target_feature = "avx2")]
+    println!("AVX2 enabled");
+    #[cfg(not(target_feature = "avx2"))]
+    println!("AVX2 disabled");
+
+    let args: Vec<String> = env::args().collect();
+    if args.len() < 4 {
+        println!("Usage: <function name> <function params> <target selector>");
+        process::exit(-1);
+    }
+
+    // remove any leading 0x
+    let selector = args[3].trim_start_matches("0x");
+    let selector = u32::from_str_radix(selector, 16)
+        .expect("Invalid number")
+        .to_be();
+    let function_name = SmallString::new(&args[1]);
+    let function_params = SmallString::new(&args[2]);
+
+    if function_name.length + function_params.length >= 115 {
+        println!("Total length of <function name> and <function params> must be under 115 bytes.");
+        process::exit(-1);
+    }
+
+    if std::mem::size_of::<u64>() != 8 {
+        println!("Incompatible architecture");
+        println!("u64: {}", std::mem::size_of::<u64>());
+        process::exit(-1);
+    }
+
+    println!("Function name: {}", args[1]);
+    println!("Function params: {}", args[2]);
+    println!("Target selector: {selector:X?}",);
+
+    let num_threads = num_cpus::get();
```

Comment:
Somehow this gives 35 instead of 36 on my system.

Reply:
that's a bit weird -- seems like maybe a bug in `num_cpus`?

correctly returns 32 on my system

Reply:
how do you feel about making the number of threads an optional param instead of hardcoding +1?

Reply:
I think let's make it optional as a 4th argument. If none is provided, use `num_cpus()`.

Reply:
done -- running it with 64 threads on my machine actually substantially increases performance

Reply:
may have found a bug though -- not finding matches for some num_threads values

Reply:
still seeing the above behavior.  for example

`foo foo 0 64` fails to find a result that `foo foo 0 32` can find in 1.5s .... I don't see anything obviously wrong, but probably worth tracking this down :thinking:

Reply:
I'm actually wondering if, for values of `num_threads` larger than the number of CPUs, there isn't some weirdness happening between optimizations and the `unsafe` intrinsics.  I'm unable to get it to misbehave with `num_threads` <= number of cores